src/service.rs

```diff
@@ -780,23 +778,58 @@ impl Service {
                         _ => false,
                     };
 
-                    if should_count && socket.is_ipv4() {
-                        let local_socket = self.local_enr.read().udp4_socket().map(SocketAddr::V4);
+                    if should_count {
+                        // get the advertised local addresses
+                        let (local_ip4_socket, local_ip6_socket) = {
+                            let local_enr = self.local_enr.read();
+                            (local_enr.udp4_socket(), local_enr.udp6_socket())
+                        };
+
                         if let Some(ref mut ip_votes) = self.ip_votes {
                             ip_votes.insert(node_id, socket);
-                            if let Some(majority_socket) = ip_votes.majority() {
-                                if Some(majority_socket) != local_socket {
-                                    info!("Local UDP socket updated to: {}", majority_socket);
-                                    self.send_event(Discv5Event::SocketUpdated(majority_socket));
-                                    // Update the UDP socket
-                                    if self
+                            let (maybe_ip4_majority, maybe_ip6_majority) = ip_votes.majority();
```

Comment:
There is some interesting logic here. We are not checking our `IpMode`. So if we are in Ipv4 mode, we could update our ENR to include ipv6 if enough nodes tell us this is what they are contacting us on. 

Initially, I thought we wouldn't want to do this, but perhaps we do. We could be behind a NAT that is contacting via ipv6 and sending us internally ipv4 packets. This kind of setup, is going to be tricky as to what we allow in our routing table.

Probably should have a chat about the kind of logic we're going to support here (and make a doc). If we are in ipv4 mode, should we advertise ipv6 and vice-versa?
If we allow ipv4-only and ipv6-only to advertise their counter parts, then what should each allow in our local routing table. If we are dual stack, should we preference ipv6? If so, there could be an attack where people force an ipv4-only node to advertise an ipv6 address (erroneously), which will make them instantly non-contactable by nodes that prefer ipv6 when both are advertised.

Maybe the simplest approach is to not support networks where nodes are ipv4 local but NAT'd behind ipv6. Curious about your thoughts here.

Reply:
So basically the thing is that the type of socket we use, and therefor the type of nodes we are able to send packets to, it's in principle not related to the address in which we are contactable. So for example, with this, now an ipv4 only node would be able to advertise itself to ipv6 (including dual stack) via tunneling